merlin/core/protocols.py

```diff
@@ -0,0 +1,116 @@
+#
+# Copyright (c) 2022, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from typing import Protocol, runtime_checkable
+
+
+@runtime_checkable
+class DictLike(Protocol):
+    def __iter__(self):
+        return iter({})
+
+    def __len__(self):
+        return 0
+
+    def __getitem__(self, key):
+        ...
+
+    def __setitem__(self, key, value):
+        ...
+
+    def __delitem__(self, key):
+        ...
+
+    def keys(self):
+        ...
+
+    def items(self):
+        ...
+
+    def values(self):
+        ...
+
+    def update(self, other):
+        ...
+
+    def copy(self):
+        ...
+
+
+@runtime_checkable
+class SeriesLike(Protocol):
+    def values(self):
+        ...
+
+    def dtype(self):
+        ...
+
+    def __getitem__(self, index):
+        ...
+
+    def __eq__(self, other):
+        ...
+
+
+@runtime_checkable
+class Transformable(DictLike, Protocol):
```

Comment:
In addition to the methods dictionary methods that are shared by dataframes, there are a few methods from dataframes that we use so frequently that it's easier to wrap a dictionary in a class and add them to the wrapper class than it would be to refactor the whole code base to do without them.

Reply:
I'm wondering if the wrapper class is an implementation detail of the `LocalExecutor`, then it could potentially be encapsulated inside it so that regular dicts could be passed? Would that be a useful thing to do for the use-case in systems for example?

Reply:
It's possible that we could do that, but it seems like a potential instance of the [primitive obsession](https://wiki.c2.com/?PrimitiveObsession) code smell and would deprive us a place to add methods with additional functionality related to e.g. transferring data across frameworks or treating multiple arrays as a single column. 🤔